core/src/pow/types.rs

```diff
@@ -391,21 +391,54 @@ impl Proof {
 	}
 }
 
-#[inline(always)]
+fn extract_bits(bits: &Vec<u8>, bit_start: usize, bit_count: usize, read_from: usize) -> u64 {
+	let mut buf: [u8; 8] = [0; 8];
+	// read 8 bytes
+	buf.copy_from_slice(&bits[read_from..read_from + 8]);
+	// convert litl endian bits to u64
+	let mut nonce = u64::from_le_bytes(buf);
+	// shake off high bits (they are part of prev number)
+	nonce <<= 64 - (bit_start - read_from * 8) - bit_count;
+	// shake of low bits (next number if any)
+	nonce >> 64 - bit_count
```

Comment:
Unwanted bits should be masked out rather than shifted out on the left:
let skip_bits = bit_start - read_from * 8;
let bit_mask = (1 << bit_count) - 1;
u64::from_le_bytes(buf) >> skip_bits & bit_mask;

Reply:
applied the change, but we need an additional check -  if `bit_count == 64` the second line will panic